lang/rust/src/decimal.rs

```diff
@@ -45,30 +67,271 @@ impl Decimal {
         let sign_byte = 0xFF * u8::from(self.value.sign() == Sign::Minus);
         let mut decimal_bytes = vec![sign_byte; len];
         let raw_bytes = self.value.to_signed_bytes_be();
-        let num_raw_bytes = raw_bytes.len();
-        let start_byte_index = len.checked_sub(num_raw_bytes).ok_or(Error::SignExtend {
+        let raw_bytes_len = raw_bytes.len();
+        let start_byte_index = len.checked_sub(raw_bytes_len).ok_or(Error::SignExtend {
             requested: len,
-            needed: num_raw_bytes,
+            needed: raw_bytes_len,
         })?;
         decimal_bytes[start_byte_index..].copy_from_slice(&raw_bytes);
         Ok(decimal_bytes)
     }
+
+    pub fn from_f64(decimal: f64) -> AvroResult<Self> {
+        if decimal.is_nan() {
+            return Err(Error::DecimalPrecisionNAN);
+        }
+
+        let as_str = format!("{}", decimal);
+        let (decimal_str, precision, scale) = match as_str.find('.') {
+            Some(index) => {
+                let lhs = as_str.get(..index).unwrap();
+                let rhs = as_str.get(index + 1..).unwrap();
+                let number = format!("{}{}", lhs, rhs);
+                let rhs_len = rhs.len();
+                (number, lhs.len() + rhs_len, rhs_len)
+            }
+            None => (as_str.clone(), as_str.len(), 0),
+        };
+
+        // f64::MAX is 309 digits, so we can't have more than 308.
+        if precision > 308 {
+            return Err(Error::DecimalPrecisionTooLarge { precision });
+        }
+
+        Ok(Self {
+            value: BigInt::from_str(decimal_str.as_str()).unwrap(),
+            len: decimal_str.len(),
+            precision,
+            scale,
+        })
+    }
+
+    /// Create a new decimal from a signed Big-Endian encoded byte array
+    pub fn from_bytes<T: AsRef<[u8]>>(
+        bytes: T,
+        precision: usize,
+        scale: usize,
+    ) -> AvroResult<Self> {
+        let bytes_ref = bytes.as_ref();
+        let big_int = BigInt::from_signed_bytes_be(bytes_ref);
+
+        let as_string = big_int.to_string();
```

Comment:
This seems very inefficient. You only need to compare it to `abs(x) < BigInt(10).powi(precision + 1)` which should be much faster to compute than the printed string.